src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs

```diff
@@ -12,14 +16,154 @@ namespace System.Net.Http
 {
     internal partial class CurlHandler : HttpMessageHandler
     {
+        private sealed class ClientCertProvider : IDisposable
+        {
+            internal readonly GCHandle _gcHandle;
+            internal readonly Interop.Ssl.ClientCertCallback _callback;
+            private SafeEvpPKeyHandle _privateKeyHandle = null;
+            private SafeX509Handle _certHandle = null;
+
+            internal ClientCertProvider ()
+            {
+                _gcHandle = GCHandle.Alloc(this);
+                _callback = TlsClientCertCallback;
+            }
+
+            private int TlsClientCertCallback(IntPtr ssl, out IntPtr certHandle, out IntPtr privateKeyHandle)
+            {
+                using (SafeSslHandle sslHandle = new SafeSslHandle(ssl))
+                {
+                    certHandle = IntPtr.Zero;
+                    privateKeyHandle = IntPtr.Zero;
+                    VerboseTrace("libssl's client certificate callback");
+
+                    ISet<string> issuerNames = GetRequestCertificateAuthorities(sslHandle);
+                    X509Certificate2 certificate;
+                    X509Chain chain;
+                    if (!GetClientCertificate(issuerNames, out certificate, out chain))
+                    {
+                        VerboseTrace("no cert or chain");
+                        return 0;
+                    }
+
+                    Debug.Assert(chain != null && chain.ChainElements.Count > 0, " a X509Chain received must not be empty");
+
+                    using (RSAOpenSsl rsa = certificate.GetRSAPrivateKey() as RSAOpenSsl)
+                    using (ECDsaOpenSsl ecdsa = certificate.GetECDsaPrivateKey() as ECDsaOpenSsl)
+                    {
+                        if (rsa != null)
+                        {
+                            _privateKeyHandle = rsa.DuplicateKeyHandle();
+                        }
+                        else if (ecdsa != null)
+                        {
+                            _privateKeyHandle = ecdsa.DuplicateKeyHandle();
+                        }
+                    }
```

Comment:
While I wrote this originally, and I defended it in the last PR, I did do it as Stephen asks here when I added ECDSA to SslStream's server certs.